modules/miner/miner.go

```diff
@@ -229,6 +240,10 @@ func (m *Miner) BlocksMined() (goodBlocks, staleBlocks int) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
+	if m.tg.IsStopped() {
+		build.Critical(errClosed)
+	}
```

Comment:
You need both. First you check if it's stopped. If it is, you either return an error or call `build.Critical`. Then you do the `tg.Add(); defer tg.Done()`. You need both.